In [1]:
import tensorflow as tf
import pandas as pd
tf.__version__

'2.13.0'

In [2]:
# Read training data - table of elements separated by spaces for columns, and rows as newlines
X_train = pd.read_table("../input_data/other/shuttle.trn", header=None, sep=' ')
X_train.head()

0   1   2  3   4   5   6   7   8  9
0  50  21  77  0  28   0  27  48  22  2
1  55   0  92  0   0  26  36  92  56  4
2  53   0  82  0  52  -5  29  30   2  1
3  37   0  76  0  28  18  40  48   8  1
4  37   0  79  0  34 -26  43  46   2  1

In [3]:
# Set target to be column 9 of the dataset (and remove it from the feature set)
y_train = X_train.pop(9)
y_train.head()

0    2
1    4
2    1
3    1
4    1
Name: 9, dtype: int64

In [4]:
# Load testing data
X_test = pd.read_table("../input_data/other/shuttle.tst", header=None, sep=' ')
X_test.head()

0  1   2  3   4   5   6   7   8  9
0  55  0  81  0  -6  11  25  88  64  4
1  56  0  96  0  52  -4  40  44   4  4
2  50 -1  89 -7  50   0  39  40   2  1
3  53  9  79  0  42  -2  25  37  12  4
4  55  2  82  0  54  -6  26  28   2  1

In [5]:
# Split feature set and target
y_test = X_test.pop(9)
y_test.head()

0    4
1    4
2    1
3    4
4    1
Name: 9, dtype: int64

In [6]:
# Ensure output is reproducible
tf.random.set_seed(8)

# Create a new model
model = tf.keras.Sequential()
# 9 input layer with 512 fully connected nodes
fc1 = tf.keras.layers.Dense(512, input_shape=(9,), activation='relu')
# Fully connected hidden layers
fc2 = tf.keras.layers.Dense(512, activation='relu')
fc3 = tf.keras.layers.Dense(128, activation='relu')
fc4 = tf.keras.layers.Dense(128, activation='relu')
# Output layer for probabilities between 8 categories - Softmax activation for categorical model
fc5 = tf.keras.layers.Dense(8, activation='softmax')

In [7]:
# Add the layers to the model
model.add(fc1)
model.add(fc2)
model.add(fc3)
model.add(fc4)
model.add(fc5)
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 512)               5120      
                                                                 
 dense_1 (Dense)             (None, 512)               262656    
                                                                 
 dense_2 (Dense)             (None, 128)               65664     
                                                                 
 dense_3 (Dense)             (None, 128)               16512     
                                                                 
 dense_4 (Dense)             (None, 8)                 1032      
                                                                 
Total params: 350984 (1.34 MB)
Trainable params: 350984 (1.34 MB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [8]:
# Sparse loss function will train the model to only place its probability to one output
loss = tf.keras.losses.SparseCategoricalCrossentropy()
# Use adam optimiser function
optimizer = tf.keras.optimizers.Adam(0.001)
# Compile the model using the selected function
model.compile(optimizer=optimizer, loss=loss, metrics=['accuracy'])
# Train the model over 5 epochs
model.fit(X_train, y_train, epochs=5)

Epoch 1/5
1360/1360 [==============================] - 10s 6ms/step - loss: 0.5382 - accuracy: 0.9773
Epoch 2/5
1360/1360 [==============================] - 8s 6ms/step - loss: 0.2099 - accuracy: 0.9939
Epoch 3/5
1360/1360 [==============================] - 8s 6ms/step - loss: 0.0902 - accuracy: 0.9951
Epoch 4/5
1360/1360 [==============================] - 9s 6ms/step - loss: 0.0519 - accuracy: 0.9966
Epoch 5/5
1360/1360 [==============================] - 8s 6ms/step - loss: 0.0419 - accuracy: 0.9958


In [9]:
# Test the model on the testing dataset
model.evaluate(X_test, y_test)

454/454 [==============================] - 1s 2ms/step - loss: 0.0151 - accuracy: 0.9973


[0.015117587521672249, 0.9973103404045105]

# Accuracy in this model is 99.73% as shown by the evaluation output